crates/pathfinder/src/sync/p2p.rs

```diff
@@ -131,6 +148,158 @@ impl Sync {
+                    },
+                };
+
+                let mut responses =
+                    match self.p2p.send_transactions_sync_request(peer, request).await {
+                        Ok(x) => x,
+                        Err(error) => {
+                            // Failed to establish connection, try next peer.
+                            tracing::debug!(%peer, reason=%error, "Transactions request failed");
+                            continue 'next_peer;
+                        }
+                    };
+
+                let mut transactions = Vec::new();
+                while let Some(transaction) = responses.next().await {
+                    match transaction {
+                        TransactionsResponse::Transaction(tx) => {
+                            match Transaction::try_from_dto(tx) {
+                                Ok(tx) if transactions.len() < curr_block.transaction_count => {
+                                    transactions.push(tx)
+                                }
+                                Ok(tx) => {
+                                    if !check_transactions(&curr_block, &transactions).await? {
+                                        tracing::debug!(
+                                            "Invalid transactions for block {}, trying next peer",
+                                            curr_block.number
+                                        );
+                                        continue 'next_peer;
+                                    }
+                                    transactions::persist(
+                                        self.storage.clone(),
+                                        curr_block.clone(),
+                                        transactions.clone(),
+                                    )
+                                    .await
+                                    .context("Inserting transactions")?;
+                                    if curr_block.number == last_block {
+                                        return Ok(());
+                                    }
+                                    curr_block =
+                                        headers::query(self.storage.clone(), curr_block.number + 1)
+                                            .await?
+                                            .ok_or_else(|| {
+                                                anyhow::anyhow!("Next block not found")
+                                            })?;
+                                    transactions.clear();
+                                    transactions.push(tx);
+                                }
+                                Err(error) => {
+                                    tracing::debug!(%peer, %error, "Transaction stream returned unexpected DTO");
+                                    continue 'next_peer;
+                                }
+                            }
+                        }
```

Comment:
This arm should move outside of the match.

The final block will never complete since it needs to receive a additional transaction in order to process this arm.

Or alternatively expect some number of transactions with a double loop.
```rust
for _ in 0..header.transaction_count {
    let Some(tx) = responses.next.await else {
         // I think this means the sender closed the channel unexpectedly?
         // Should log/punish peer here.
    };
    match tx {
       ...
       match Transaction::try_from_dto(tx) {
           Ok(tx) => transactions.push(tx),
           ...
       }
    }
}

// Check transactions

// Persist

// Reset and repeat for next block
```

Reply:
Ah yeah I missed that 👍